File: src/main/java/org/radarbase/management/security/jwt/algorithm/AsymmetricalJwtAlgorithm.java
package org.radarbase.management.security.jwt.algorithm;

import java.security.KeyPair;
import java.util.Base64;

import org.radarbase.auth.jwks.JsonWebKey;
import org.radarbase.auth.jwks.MPJsonWebKey;

public abstract class AsymmetricalJwtAlgorithm implements JwtAlgorithm {

    protected final KeyPair keyPair;

    protected AsymmetricalJwtAlgorithm(KeyPair keyPair) {
        this.keyPair = keyPair;
    }

    /** Header used for encoding public keys. */
    protected abstract String getEncodedStringHeader();

    /** Footer used for encoding public keys. */
    protected abstract String getEncodedStringFooter();

    /** The family of cryptographic algorithms used with the key. */
    protected abstract String getKeyType();

    @Override
    public String getVerifierKeyEncodedString() {
        return getEncodedStringHeader() + '\n'
                + new String(Base64.getEncoder().encode(keyPair.getPublic().getEncoded()))
                + '\n' + getEncodedStringFooter();
    }

    @Override
    public JsonWebKey getJwk() {
        return new MPJsonWebKey(
                this.getAlgorithm().getName(),
                this.getKeyType(),
                this.getVerifierKeyEncodedString());
    }
}

File: src/main/java/org/radarbase/management/security/jwt/algorithm/AsymmetricalJwtAlgorithm.kt
package org.radarbase.management.security.jwt.algorithm

import org.radarbase.auth.jwks.JsonWebKey
import org.radarbase.auth.jwks.MPJsonWebKey
import java.security.KeyPair
import java.util.*

abstract class AsymmetricalJwtAlgorithm protected constructor(protected val keyPair: KeyPair) : JwtAlgorithm {
    protected abstract val encodedStringHeader: String
    protected abstract val encodedStringFooter: String
    protected abstract val keyType: String
    override val verifierKeyEncodedString: String
        get() = """
             ${encodedStringHeader}
             ${String(Base64.getEncoder().encode(keyPair.public.encoded))}
             ${encodedStringFooter}
             """.trimIndent()
    override val jwk: JsonWebKey
        get() = MPJsonWebKey(
            algorithm.name,
            keyType,
            verifierKeyEncodedString
        )
}
